File: realtime/src/zocoBridge.ts
export type TurnReply = {
  text: string;
  endCall?: boolean;
  disposition?: string | null;
  transfer?: string | null;
  provider?: string | null;
  model?: string | null;
};

export type ToolReply = {
  ok: boolean;
  result?: string;
  endCall?: boolean;
  disposition?: string | null;
  transfer?: string | null;
  say?: string;
  duplicate?: boolean;
};

export type SessionSnapshot = {
  callId: string;
  greeting: string;
  instructions?: string;
  knowledge?: string;
  language: string;
  agent: {
    id: string;
    name: string;
    language: string;
    ttsVoice?: string;
    ttsModel?: string;
    ttsProvider?: string;
    transferNumber?: string;
    callSettings?: Record<string, unknown>;
    voiceRuntime?: string;
    gender?: "male" | "female" | string;
    customTools?: Array<{ id?: string; name: string; description?: string }>;
  };
  llm: {
    provider: string;
    model: string;
    baseUrl: string;
    apiKey: string;
  } | null;
  customer: { name?: string; phone?: string };
};

export type SessionEventPayload = {
  eventId: string;
  type: "transcript" | "status" | "metric" | "recording" | "disposition";
  role?: "user" | "assistant" | "system";
  text?: string;
  status?: string;
  disposition?: string;
  reason?: string;
  metrics?: Record<string, number | string>;
  recordingUrl?: string;
  duplicate?: boolean;
};

function isPlaceholder(value: string) {
  const raw = String(value || "").trim();
  return !raw || raw === "..." || /^<.*>$/.test(raw);
}

function resolvedBridgeToken() {
  const raw = String(process.env.LIVEKIT_BRIDGE_TOKEN || "").trim();
  if (!isPlaceholder(raw)) return raw;
  const secret = String(process.env.LIVEKIT_API_SECRET || "").trim();
  if (!secret) throw new Error("LIVEKIT_BRIDGE_TOKEN is not configured");
  return `zoco-${secret.slice(0, 24)}`;
}

function bridgeConfig() {
  const baseUrl = String(process.env.ZOCO_BRIDGE_URL || process.env.PUBLIC_BASE_URL || "http://127.0.0.1:8787").replace(/\/$/, "");
  const token = resolvedBridgeToken();
  return { baseUrl, token };
}

function headers(token: string) {
  return {
    Authorization: `Bearer ${token}`,
    "Content-Type": "application/json",
  };
}

export async function fetchSnapshot(callId: string): Promise<SessionSnapshot> {
  const { baseUrl, token } = bridgeConfig();
  const response = await fetch(`${baseUrl}/api/livekit/sessions/${encodeURIComponent(callId)}/snapshot`, {
    headers: headers(token),
  });
  const data = await response.json().catch(() => ({}));
  if (!response.ok) {
    throw new Error(data.error || `Snapshot failed (${response.status})`);
  }
  return data as SessionSnapshot;
}

export async function postTurn(callId: string, payload: {
  eventId: string;
  userText: string;
  sttLanguage?: string | null;
}): Promise<TurnReply> {
  const { baseUrl, token } = bridgeConfig();
  const response = await fetch(`${baseUrl}/api/livekit/sessions/${encodeURIComponent(callId)}/turn`, {
    method: "POST",
    headers: headers(token),
    body: JSON.stringify(payload),
  });
  const data = await response.json().catch(() => ({}));
  if (!response.ok) {
    throw new Error(data.error || `Turn failed (${response.status})`);
  }
  return data as TurnReply;
}

export async function postTool(callId: string, payload: {
  eventId: string;
  name: string;
  args?: Record<string, unknown>;
}): Promise<ToolReply> {
  const { baseUrl, token } = bridgeConfig();
  const response = await fetch(`${baseUrl}/api/livekit/sessions/${encodeURIComponent(callId)}/tools`, {
    method: "POST",
    headers: headers(token),
    body: JSON.stringify(payload),
  });
  const data = await response.json().catch(() => ({}));
  if (!response.ok) {
    throw new Error(data.error || `Tool failed (${response.status})`);
  }
  return data as ToolReply;
}

export async function postEvent(callId: string, payload: SessionEventPayload): Promise<{ ok: boolean; duplicate?: boolean }> {
  const { baseUrl, token } = bridgeConfig();
  const response = await fetch(`${baseUrl}/api/livekit/sessions/${encodeURIComponent(callId)}/events`, {
    method: "POST",
    headers: headers(token),
    body: JSON.stringify(payload),
  });
  const data = await response.json().catch(() => ({}));
  if (!response.ok) {
    throw new Error(data.error || `Event failed (${response.status})`);
  }
  return data as { ok: boolean; duplicate?: boolean };
}

File: pipecat/zoco_bridge.py
import asyncio
import os
import uuid
from typing import Any

import aiohttp

def _is_placeholder(value: str) -> bool:
    raw = str(value or "").strip()
    return not raw or raw == "..." or (raw.startswith("<") and raw.endswith(">"))


def _bridge_token() -> str:
    raw = str(os.getenv("PIPECAT_BRIDGE_TOKEN") or os.getenv("LIVEKIT_BRIDGE_TOKEN") or "").strip()
    if not _is_placeholder(raw):
        return raw
    secret = str(os.getenv("LIVEKIT_API_SECRET") or os.getenv("PIPECAT_API_SECRET") or "").strip()
    if not secret:
        raise RuntimeError("PIPECAT_BRIDGE_TOKEN is not configured")
    return f"zoco-{secret[:24]}"


def _base_url() -> str:
    return str(os.getenv("ZOCO_BRIDGE_URL") or os.getenv("PUBLIC_BASE_URL") or "http://127.0.0.1:8787").rstrip("/")


def event_id(prefix: str) -> str:
    return f"{prefix}_{uuid.uuid4().hex[:12]}"


def _headers() -> dict[str, str]:
    return {
        "Authorization": f"Bearer {_bridge_token()}",
        "Content-Type": "application/json",
    }


async def _request(method: str, path: str, payload: dict | None = None) -> dict[str, Any]:
    url = f"{_base_url()}{path}"
    async with aiohttp.ClientSession() as session:
        async with session.request(method, url, json=payload, headers=_headers()) as response:
            data = await response.json(content_type=None)
            if response.status >= 400:
                message = data.get("error") if isinstance(data, dict) else None
                raise RuntimeError(message or f"Zoco bridge failed ({response.status})")
            return data if isinstance(data, dict) else {}


async def fetch_snapshot(call_id: str) -> dict[str, Any]:
    return await _request("GET", f"/api/pipecat/sessions/{call_id}/snapshot")


async def post_tool(call_id: str, name: str, args: dict | None = None) -> dict[str, Any]:
    return await _request(
        "POST",
        f"/api/pipecat/sessions/{call_id}/tools",
        {"eventId": event_id("tool"), "name": name, "args": args or {}},
    )


async def post_event(call_id: str, payload: dict[str, Any]) -> dict[str, Any]:
    return await _request("POST", f"/api/pipecat/sessions/{call_id}/events", payload)


async def record_transcript(call_id: str, role: str, text: str) -> None:
    trimmed = str(text or "").strip()
    if not trimmed:
        return
    await post_event(
        call_id,
        {
            "eventId": event_id("usr" if role == "user" else "asst" if role == "assistant" else "sys"),
            "type": "transcript",
            "role": role,
            "text": trimmed,
        },
    )


async def record_status(call_id: str, status: str, reason: str | None = None) -> None:
    await post_event(
        call_id,
        {
            "eventId": event_id("status"),
            "type": "status",
            "status": status,
            "reason": reason,
        },
    )


async def record_disposition(call_id: str, disposition: str, reason: str | None = None) -> None:
    await post_event(
        call_id,
        {
            "eventId": event_id("disp"),
            "type": "disposition",
            "disposition": disposition,
            "reason": reason,
        },
    )


async def record_metric(call_id: str, name: str, value: float | str) -> None:
    await post_event(
        call_id,
        {
            "eventId": event_id("metric"),
            "type": "metric",
            "metrics": {"name": name, "value": value},
        },
    )


def fire(coro) -> None:
    asyncio.create_task(coro)
